Đọc file 'healthcare-dataset-stroke-data.csv' bằng thư viện pandas

In [53]:
import pandas as pd
df = pd.read_csv(r'C:\GIT\GROUP__3\DATA\healthcare-dataset-stroke-data.csv')
df.head()

id  gender   age  hypertension  heart_disease ever_married  \
0   9046    Male  67.0             0              1          Yes   
1  51676  Female  61.0             0              0          Yes   
2  31112    Male  80.0             0              1          Yes   
3  60182  Female  49.0             0              0          Yes   
4   1665  Female  79.0             1              0          Yes   

       work_type Residence_type  avg_glucose_level   bmi   smoking_status  \
0        Private          Urban             228.69  36.6  formerly smoked   
1  Self-employed          Rural             202.21   NaN     never smoked   
2        Private          Rural             105.92  32.5     never smoked   
3        Private          Urban             171.23  34.4           smokes   
4  Self-employed          Rural             174.12  24.0     never smoked   

   stroke  
0       1  
1       1  
2       1  
3       1  
4       1

**Giải thích ý nghĩa các cột trong tập dữ liệu**

| Tên cột              | Ý nghĩa |
|----------------------|--------|
| `id`                 | Mã định danh duy nhất của mỗi cá nhân |
| `gender`             | Giới tính của cá nhân (`Male`, `Female`, hoặc `Other`) |
| `age`                | Tuổi của cá nhân (tính theo năm) |
| `hypertension`       | Tăng huyết áp (0 = Không, 1 = Có) |
| `heart_disease`      | Bệnh tim (0 = Không, 1 = Có) |
| `ever_married`       | Đã từng kết hôn hay chưa (`Yes`, `No`) |
| `work_type`          | Loại hình công việc (`Private`, `Self-employed`, `Govt_job`, `children`, `Never_worked`) |
| `Residence_type`     | Loại nơi cư trú (`Urban` = thành thị, `Rural` = nông thôn) |
| `avg_glucose_level`  | Mức đường huyết trung bình (mg/dL) |
| `bmi`                | Chỉ số khối cơ thể (Body Mass Index) |
| `smoking_status`     | Tình trạng hút thuốc (`formerly smoked`, `never smoked`, `smokes`, `Unknown`) |
| `stroke`             | Đã từng bị đột quỵ hay chưa (0 = Không, 1 = Có) |


# 1. PHÂN TÍCH SƠ BỘ DỮ LIỆU

Hiểu cấu trúc dữ liệu

In [5]:
# Liệt kê các cột và xem kiểu dữ liệu 
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5110 entries, 0 to 5109
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   id                 5110 non-null   int64  
 1   gender             5110 non-null   object 
 2   age                5110 non-null   float64
 3   hypertension       5110 non-null   int64  
 4   heart_disease      5110 non-null   int64  
 5   ever_married       5110 non-null   object 
 6   work_type          5110 non-null   object 
 7   Residence_type     5110 non-null   object 
 8   avg_glucose_level  5110 non-null   float64
 9   bmi                4909 non-null   float64
 10  smoking_status     5110 non-null   object 
 11  stroke             5110 non-null   int64  
dtypes: float64(3), int64(4), object(5)
memory usage: 479.2+ KB
None


In [6]:
# thống kê mô tả bộ dữ liệu dạng số
print(df.describe())

                 id          age  hypertension  heart_disease  \
count   5110.000000  5110.000000   5110.000000    5110.000000   
mean   36517.829354    43.226614      0.097456       0.054012   
std    21161.721625    22.612647      0.296607       0.226063   
min       67.000000     0.080000      0.000000       0.000000   
25%    17741.250000    25.000000      0.000000       0.000000   
50%    36932.000000    45.000000      0.000000       0.000000   
75%    54682.000000    61.000000      0.000000       0.000000   
max    72940.000000    82.000000      1.000000       1.000000   

       avg_glucose_level          bmi       stroke  
count        5110.000000  4909.000000  5110.000000  
mean          106.147677    28.893237     0.048728  
std            45.283560     7.854067     0.215320  
min            55.120000    10.300000     0.000000  
25%            77.245000    23.500000     0.000000  
50%            91.885000    28.100000     0.000000  
75%           114.090000    33.100000     0

## Tổng quan thống kê mô tả

Bảng thống kê mô tả dưới đây phản ánh các đặc trưng cơ bản của bộ dữ liệu bệnh nhân, bao gồm 7 biến định lượng và 3 biến nhị phân. Dưới đây là phân tích chi tiết:

### Biến định danh (id)

Không có ý nghĩa phân tích thống kê, được sử dụng để định danh từng cá thể.

Phân bố ID trải đều từ 67 đến 72940, không cần xử lý thêm.

### Tuổi (age)

- Trung bình: 43.23 tuổi  
- Độ lệch chuẩn (std): 22.61  
- Giá trị nhỏ nhất: 0.08 tuổi (khoảng 1 tháng) — cần kiểm tra xem có phải dữ liệu bất thường không  
- Phân vị:
  - 25% ≤ 25 tuổi
  - 75% ≤ 61 tuổi  
  → Phân bố có vẻ lệch phải (phần lớn là người trẻ)

### Tăng huyết áp (hypertension) & Bệnh tim (heart_disease)

- Là các biến nhị phân (0/1)
- Tỷ lệ người bị tăng huyết áp: ~9.75%  
- Tỷ lệ người mắc bệnh tim: ~5.4%  
→ Tỷ lệ này khá thấp, phân bố mất cân bằng, cần lưu ý nếu xây dựng mô hình học máy

### Mức đường huyết trung bình (avg_glucose_level)

- Trung bình: 106.15 mg/dL  
- Độ lệch chuẩn: 45.28  
- Giá trị tối đa: 271.74 — có thể là outlier  
- 75% dữ liệu ≤ 114.09  
→ Phân bố có dấu hiệu lệch phải, nên chuẩn hóa hoặc biến đổi log nếu dùng trong mô hình

### Chỉ số BMI (bmi)

- Trung bình: 28.89 — nằm trong ngưỡng tiền béo phì theo chuẩn WHO  
- Số lượng quan sát hợp lệ: 4909/5110 → khoảng 3.9% thiếu dữ liệu  
- Giá trị tối đa: 97.6 — cao bất thường, khả năng là outlier  
→ Cần xử lý missing values và kiểm tra outliers

### Đột quỵ (stroke)

- Biến mục tiêu nhị phân  
- Tỷ lệ đột quỵ: 4.87%  
→ Class imbalance nghiêm trọng, cần các kỹ thuật xử lý như oversampling, SMOTE, hoặc điều chỉnh trọng số trong mô hình

## Kết luận

- **Kiểm tra outliers**: đặc biệt ở các biến `age`, `avg_glucose_level`, `bmi`
- **Xử lý missing values**: biến `bmi` cần bổ sung hoặc ước lượng dữ liệu thiếu
- **Xử lý class imbalance**: biến mục tiêu `stroke` cần được cân bằng bằng các kỹ thuật phù hợp
- **Chuẩn hóa dữ liệu**: nên thực hiện chuẩn hóa hoặc feature scaling với các biến định lượng để đảm bảo hiệu quả mô hình hóa

In [7]:
# thống kê mô tả cho các cột kiểu object
df_catogery = df.describe(include=[object])
df_catogery

gender ever_married work_type Residence_type smoking_status
count     5110         5110      5110           5110           5110
unique       3            2         5              2              4
top     Female          Yes   Private          Urban   never smoked
freq      2994         3353      2925           2596           1892

In [8]:
df['gender'].unique()

array(['Male', 'Female', 'Other'], dtype=object)

In [9]:
df.nunique()

id                   5110
gender                  3
age                   104
hypertension            2
heart_disease           2
ever_married            2
work_type               5
Residence_type          2
avg_glucose_level    3979
bmi                   418
smoking_status          4
stroke                  2
dtype: int64

In [10]:
df['smoking_status'].unique()

array(['formerly smoked', 'never smoked', 'smokes', 'Unknown'],
      dtype=object)

In [11]:
gender = df['gender'].value_counts().reset_index(name = "count")
gender

gender  count
0  Female   2994
1    Male   2115
2   Other      1

In [12]:
ever_married = df['ever_married'].value_counts().reset_index(name = "Số lượng")
ever_married

ever_married  Số lượng
0          Yes      3353
1           No      1757

In [13]:
work_type = df['work_type'].value_counts().reset_index(name = "Số lượng")
work_type

work_type  Số lượng
0        Private      2925
1  Self-employed       819
2       children       687
3       Govt_job       657
4   Never_worked        22

In [14]:
Residence_type = df['Residence_type'].value_counts().reset_index(name = "Số lượng")
Residence_type

Residence_type  Số lượng
0          Urban      2596
1          Rural      2514

In [15]:
smoking_status = df['smoking_status'].value_counts().reset_index(name = "Số lượng")
smoking_status

smoking_status  Số lượng
0     never smoked      1892
1          Unknown      1544
2  formerly smoked       885
3           smokes       789

In [29]:
# lọc lấy nhữung người trên 18 tuổi
df_18 = df.copy()
df_18 = df_18[df_18['age'] > 18]
df_18.head(10)

id  gender   age  hypertension  heart_disease ever_married  \
0   9046    Male  67.0             0              1          Yes   
1  51676  Female  61.0             0              0          Yes   
2  31112    Male  80.0             0              1          Yes   
3  60182  Female  49.0             0              0          Yes   
4   1665  Female  79.0             1              0          Yes   
5  56669    Male  81.0             0              0          Yes   
6  53882    Male  74.0             1              1          Yes   
7  10434  Female  69.0             0              0           No   
8  27419  Female  59.0             0              0          Yes   
9  60491  Female  78.0             0              0          Yes   

       work_type Residence_type  avg_glucose_level   bmi   smoking_status  \
0        Private          Urban             228.69  36.6  formerly smoked   
1  Self-employed          Rural             202.21   NaN     never smoked   
2        Private          Rural             105.92  32.5     never smoked   
3        Private          Urban             171.23  34.4           smokes   
4  Self-employed          Rural             174.12  24.0     never smoked   
5        Private          Urban             186.21  29.0  formerly smoked   
6        Private          Rural              70.09  27.4     never smoked   
7        Private          Urban              94.39  22.8     never smoked   
8        Private          Rural              76.15   NaN          Unknown   
9        Private          Urban              58.57  24.2          Unknown   

   stroke  
0       1  
1       1  
2       1  
3       1  
4       1  
5       1  
6       1  
7       1  
8       1  
9       1

In [32]:
df_18.shape

(4194, 12)

In [40]:
# ép cột 'age' từ float sang int
df['age'] = df['age'].astype(int)

In [42]:
df['age'].nunique()

83

In [30]:
# Tính phân phối giữa 2 cột 'age' và 'smoking_status'
# nhóm tuổi và tình trạng hút thuốc
age_smoking = df_18.groupby(['age', 'smoking_status']).size().reset_index(name='count')
age_smoking.sort_values(by='age', inplace=True, ascending = False)
age_smoking.head(20)

age   smoking_status  count
252  82.0          Unknown      9
253  82.0  formerly smoked     19
254  82.0     never smoked     22
255  82.0           smokes      6
248  81.0          Unknown     13
249  81.0  formerly smoked     12
250  81.0     never smoked     30
251  81.0           smokes      5
244  80.0          Unknown     13
245  80.0  formerly smoked     17
246  80.0     never smoked     33
247  80.0           smokes      7
240  79.0          Unknown     17
241  79.0  formerly smoked     20
242  79.0     never smoked     39
243  79.0           smokes      9
236  78.0          Unknown     25
237  78.0  formerly smoked     31
238  78.0     never smoked     39
239  78.0           smokes      7

In [ ]:
# Tạo bins tuổi
bins = [0, 18, 50, 82]
labels = ['child', 'adult', 'senior']

df_known = df[df['smoking_status'] != 'Unknown'].copy()
df_known['age_bin'] = pd.cut(df_known['age'], bins=bins, labels=labels, right=True)

# Xem phân bố từng nhóm trong mỗi age bin
pd.crosstab(df_known['age_bin'], df_known['smoking_status'], normalize='index') * 100

# Tạo age_bin cho nhóm Unknown
df_unknown = df[df['smoking_status'] == 'Unknown'].copy()
df_unknown['age_bin'] = pd.cut(df_unknown['age'], bins=bins, labels=labels, right=True)

import numpy as np

def assign_smoking_status(row):
    age_bin = row['age_bin']
    if age_bin == 'child':
        return np.random.choice(['never smoked', 'smokes', 'formerly smoked'], p=[0.95, 0.04, 0.01])
    elif age_bin == 'adult':
        return np.random.choice(['never smoked', 'smokes', 'formerly smoked'], p=[0.35, 0.5, 0.15])
    elif age_bin == 'senior':
        return np.random.choice(['never smoked', 'smokes', 'formerly smoked'], p=[0.2, 0.1, 0.7])
    else:
        return 'never smoked'  # fallback

# Áp dụng gán
df_unknown['smoking_status_imputed'] = df_unknown.apply(assign_smoking_status, axis=1)

In [ ]:
# Cập nhật giá trị suy đoán
df.loc[df['smoking_status'] == 'Unknown', 'smoking_status'] = df_unknown['smoking_status_imputed']


In [49]:
smoking_status = df['smoking_status'].value_counts().reset_index(name='Số lượng')

smoking_status

smoking_status  Số lượng
0     never smoked      2794
1  formerly smoked      1256
2           smokes      1060

In [55]:
df['smoking_status'].unique()

array(['formerly smoked', 'never smoked', 'smokes', 'Unknown'],
      dtype=object)

Phân tích giá trị thiếu

In [16]:
# kiểm tra giá trị thiếu
df.isnull().sum()

id                     0
gender                 0
age                    0
hypertension           0
heart_disease          0
ever_married           0
work_type              0
Residence_type         0
avg_glucose_level      0
bmi                  201
smoking_status         0
stroke                 0
dtype: int64

Hầu hết các trường đều không thiếu chỉ có cột bmi là thiếu 201 giá trị tương ứng với khoảng 4% so với tổng thể.

In [ ]:
# tính tỷ lệ phần trăm thiếu
missing = df.isnull().mean() * 100
missing

In [ ]:
# Visualize giá trị thiếu bằng biểu đồ
import matplotlib.pyplot as plt
import seaborn as sns
plt.figure(figsize=(12, 6))
sns.barplot(x=missing.index, y=missing.values)
plt.title('Tỷ lệ phần trăm giá trị thiếu trong các cột')
plt.xlabel('Cột')
plt.ylabel('Tỷ lệ phần trăm (%)')
plt.xticks(rotation=30)
plt.show()

Phân tích giá trị ngoại lai

In [ ]:
# Lọc các cột số
numeric_cols = df.select_dtypes(include=['float64', 'int64']).columns

# Hàm xác định outliers theo IQR
def detect_outliers_iqr(df, cols):
    outlier_dict = {}
    for col in cols:
        Q1 = df[col].quantile(0.25)
        Q3 = df[col].quantile(0.75)
        IQR = Q3 - Q1
        lower_bound = Q1 - 1.5 * IQR
        upper_bound = Q3 + 1.5 * IQR
        outliers = df[(df[col] < lower_bound) | (df[col] > upper_bound)]
        outlier_dict[col] = {
            "số_lượng_ngoại_lai": outliers.shape[0],
            "tỷ_lệ_ngoại_lai": round(100 * outliers.shape[0] / df.shape[0], 2)
        }
    return pd.DataFrame(outlier_dict).T

# Áp dụng
outlier_summary = detect_outliers_iqr(df, numeric_cols)
print(outlier_summary)

In [ ]:
plt.figure(figsize=(10, 6))
sns.barplot(
    x=outlier_summary.index,
    y='tỷ_lệ_ngoại_lai',
    data=outlier_summary,
    palette='Blues_r'
)
plt.xticks(rotation=45)
plt.title('Tỷ lệ phần trăm giá trị ngoại lai theo từng cột số')
plt.ylabel('Tỷ lệ ngoại lai (%)')
plt.xlabel('Tên cột')
plt.tight_layout()
plt.show()

Phân tích mối quan hệ giữa các biến

In [ ]:
# Tính hệ số tương quan Pearson giữa các cột số
correlation_matrix = df[numeric_cols].corr(method='pearson')

plt.figure(figsize=(10, 8))
sns.heatmap(
    correlation_matrix,
    annot=True,             # hiển thị giá trị số
    fmt=".2f",              # định dạng số
    cmap='coolwarm',        # màu sắc
    linewidths=0.5,
    square=True,
    cbar_kws={"shrink": .75}
)
plt.title("Ma trận tương quan giữa các biến số", fontsize=14)
plt.xticks(rotation=45)
plt.yticks(rotation=0)
plt.tight_layout()
plt.show()

## Phân tích Ma trận Tương quan

### Tổng quan

Ma trận thể hiện hệ số tương quan Pearson giữa các biến định lượng và nhị phân. Giá trị dao động từ -1 đến 1, trong đó:

- +1: tương quan tuyến tính dương hoàn hảo  
- 0: không có tương quan tuyến tính  
- -1: tương quan tuyến tính âm hoàn hảo  

### Những phát hiện chính

#### Tương quan với biến mục tiêu (`stroke`)

Mọi biến đều có tương quan rất yếu với `stroke`:

- `age`: 0.25 → tương quan dương nhẹ, người lớn tuổi dễ bị đột quỵ hơn  
- `hypertension`, `heart_disease`, `avg_glucose_level`: khoảng 0.13 → có ý nghĩa lâm sàng nhưng rất yếu về mặt thống kê tuyến tính  
- `bmi`: 0.04 → gần như không liên quan

**Nhận định**: Biến mục tiêu `stroke` không có mối liên hệ tuyến tính mạnh với bất kỳ biến nào.  
→ Cần áp dụng mô hình phi tuyến (như tree-based models, XGBoost) hoặc phân tích thêm các mối liên hệ phi tuyến (mutual information, ANOVA,...)

#### Tương quan giữa các biến độc lập

- `age` có tương quan nhẹ với:
  - `bmi`: 0.33  
  - `hypertension`: 0.28  
  - `heart_disease`: 0.26  
  - `avg_glucose_level`: 0.24  

Những mối quan hệ này phản ánh đặc điểm thực tế: tuổi cao thường đi kèm với nguy cơ tăng huyết áp, bệnh tim, béo phì và đường huyết cao.

- Tương quan giữa `hypertension`, `heart_disease`, `avg_glucose_level` đều < 0.2  
→ Các biến tương đối độc lập với nhau, thuận lợi cho mô hình hóa do ít đa cộng tuyến

#### Biến không mang ý nghĩa phân tích

- `id` có tương quan gần như bằng 0 với mọi biến  
→ Là biến định danh, nên loại bỏ khỏi quá trình phân tích

### Kết luận và Khuyến nghị

- Không có biến nào có tương quan tuyến tính mạnh với `stroke`  
→ Cần tiếp cận đa chiều hơn, sử dụng các kỹ thuật khai thác đặc trưng phi tuyến hoặc kết hợp nhiều đặc trưng

- Không có dấu hiệu đa cộng tuyến nghiêm trọng giữa các biến độc lập  
→ Thuận lợi cho việc xây dựng mô hình tuyến tính

### Đề xuất

- Khám phá thêm các mối quan hệ phi tuyến hoặc tương tác biến  
- Áp dụng phương pháp chọn biến (feature selection) như:
  - RFE (Recursive Feature Elimination)  
  - `mutual_info_classif`  
  - Mô hình Random Forest và đánh giá `feature_importance`


Phân tích biến mục tiêu

In [ ]:
stroke = df['stroke'].value_counts().reset_index(name = "Số lượng")
stroke

In [ ]:
plt.figure(figsize=(8, 5))
sns.countplot(x='stroke', data=df, palette='Set2')
plt.title('Số lượng bệnh nhân bị đột quỵ và không bị đột quỵ')
plt.xlabel('Bị đột quỵ (1) / Không bị đột quỵ (0)')
plt.ylabel('Số lượng')
plt.xticks([0, 1], ['Không bị đột quỵ', 'Bị đột quỵ'])
plt.show()

Kiểm tra giá trị trùng lặp

In [ ]:
# kiểm tra trùng lặp 
duplicates = df.duplicated().sum()
print(f'Số lượng dòng trùng lặp: {duplicates}')

# XỬ LÝ DỮ LIỆU

TIỀN XỬ LÝ DỮ LIỆU

In [ ]:
# điền giá trị thiếu cho cột 'bmi'
df['bmi'].fillna(df['bmi'].mean(), inplace=True)

In [ ]:
df.isnull().sum()

In [ ]:
# xóa cột id vì không có ý nghĩa trong phân tích
df.drop(columns=['id'], inplace=True)

In [ ]:
df.head()

In [ ]:
# xử lý giá trị ngoại lai
'''Các cột như hypertension, heart_disease, stroke là biến nhị phân (0 hoặc 1) → không nên xử lý ngoại lai bằng IQR vì 1 là giá trị hợp lệ.'''

In [ ]:
def log_clip_column(df, col):
    # Log transform
    log_col = np.log1p(df[col])
    
    # Tính IQR trên biến log
    Q1 = log_col.quantile(0.25)
    Q3 = log_col.quantile(0.75)
    IQR = Q3 - Q1
    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR
    
    # Clip(cắt) ngoại lai trên biến log
    log_col_clipped = log_col.clip(lower=lower_bound, upper=upper_bound)
    
    return log_col_clipped

# Áp dụng cho cả 2 cột
df['avg_glucose_level_log'] = log_clip_column(df, 'avg_glucose_level')
df['bmi_log'] = log_clip_column(df, 'bmi')

In [ ]:
# Kiểm tra lại outliers sau khi xử lý

numeric_cols = df.select_dtypes(include=['float64', 'int64']).columns

# gọi lại hàm kiểm tra outliers
outlier_summary_post = detect_outliers_iqr(df, numeric_cols)
print(outlier_summary_post)

In [ ]:
# xóa giá trị other trong cột gender
df = df[df['gender'] != 'Other'].reset_index(drop=True)

In [ ]:
df['gender'].unique()

In [ ]:
# các cột phân loại cần mã hóa:
categorical_cols = ['gender', 'work_type', 'Residence_type', 'smoking_status']

# Áp dụng one-hot encoding
df_encoded = pd.get_dummies(df, columns=categorical_cols, drop_first=True)

### Ưu và nhược điểm của One-Hot Encoding

**Ưu điểm:**
- Không tạo quan hệ thứ tự giả giữa các giá trị phân loại.
- Dễ hiểu, trực quan và phổ biến.
- Phù hợp với nhiều thuật toán học máy.
- Thực hiện nhanh, có sẵn trong pandas và sklearn.

**Nhược điểm:**
- Tăng số chiều dữ liệu, gây tốn bộ nhớ.
- Không hiệu quả với biến phân loại có nhiều giá trị (high cardinality).
- Không thể hiện mối quan hệ gần nhau giữa các nhãn.
- Có thể làm giảm hiệu suất với một số thuật toán do dữ liệu rất nhiều chiều.


In [ ]:
# chuẩn hóa cột 'ever_married' về dạng nhị phân
df_encoded['ever_married'] = df_encoded['ever_married'].map({'Yes': 1, 'No': 0})

In [ ]:
from sklearn.preprocessing import StandardScaler

numeric_cols = ['age', 'avg_glucose_level', 'bmi']

scaler = StandardScaler()
df_encoded[numeric_cols] = scaler.fit_transform(df_encoded[numeric_cols])

In [ ]:
df_encoded.head()

**Lý do nên dùng StandardScaler**

- **Chuẩn hóa dữ liệu về phân phối chuẩn (mean=0, std=1)** giúp nhiều thuật toán học máy hoạt động hiệu quả hơn (ví dụ: Logistic Regression, SVM, Neural Networks).
- StandardScaler xử lý tốt dữ liệu có phân phối gần chuẩn, làm cho các đặc trưng có cùng thang đo, tránh việc đặc trưng có giá trị lớn chi phối mô hình.
- Các phương pháp khác như MinMaxScaler nhạy cảm với ngoại lai, còn RobustScaler thích hợp khi dữ liệu có nhiều ngoại lai nhưng không phải lúc nào cũng cần thiết.
- Do đó, StandardScaler là lựa chọn phổ biến, phù hợp với nhiều bài toán và mô hình học máy.